Vault: Add Okta to agentic IAM (#3241)
hashicorp/web-unified-docs: content/vault/v2.x/content/ai/iam/providers/okta.mdx

---
layout: docs
page_title: Configure Okta for agent IAM
description: >-
  Use Okta and Vault to secure agent permissions with the agent registry and OAuth resource server profiles.
---

# Configure Okta for agent IAM

> @include 'alerts/beta.mdx'

Reviewer: ```suggestion
```
The 2.1 release is GA for agent IAM

> @include 'alerts/enterprise-and-hcp.mdx'

Reviewer: Are we sure the feature is supported in HVD?

Reviewer 2: Mohan requested us to mention that HVD supports this as well.

Configure Okta to use with Vault Enterprise agentic IAM. Follow these workflows
to configure Okta for your agents or other applications to use Okta-issued
credentials with either authorization code or client credentials grant types.

> These workflows cover agent a

Reviewer: ```suggestion
```

## Before you start

- **You must have an
  [Okta Integrator](https://developer.okta.com/signup/) account** to configure
  the authorization server for agents.

- **You must have an [Okta Platform](https://developer.okta.com/signup/)
  account** with the non-human identities (NHI) feature available to generate
  client credentials for agents. This feature is not available in the free
  integrator account trial.

- **You must have the [`oauth2c`](https://github.com/SecureAuthCorp/oauth2c)
  CLI**. You need `oauth2c` installed and in your system `PATH` to interact
  with OAuth 2.0 authorization servers.

- **You must have [`curl`](https://curl.se/)** or some other tool to make
  API calls.

- **You must have [`jq`](https://jqlang.org/) installed** to use the example
  commands as written.

Choose to use either an authorization code or client credentials grant type
when defining your authorization server details.

<Tip>

You can use the workflow shown here to set up Okta for either authorization
code or client credentials grant types.

> </Tip>

Reviewer: ```suggestion
You can follow the directions below to use either an authorization code or client credentials grant type
when defining your authorization server details.
```
Unless I'm misunderstanding something, these two sentences are basically saying the same thing, no?

> ## Step 1: Create the OIDC Service App

Reviewer: ```suggestion
## Step 1: Create the OIDC service app
```
Style correction: use sentence case for headings. As far as I can tell, Okta doesn't use "Service App" as a proprietary phrase, so there's no need to capitalize it

You must first configure Okta with a Service App and custom authorization
server. You then define an authorization server policy and default scope

> for reading Vault secrets.

Reviewer: ```suggestion
Configure Okta with a service app and custom authorization
server, then define an authorization server policy and default scope
for reading Vault secrets.
```
Style correction: use declarative language to provide instructions

1. Within the Okta admin console, click **Applications and Resources**.

1. Click **Create App Integration**.

   ![Create new App Integration](/img/agentic-ai/screenshot-okta-app000.png)

1. Click **OIDC - OpenID Connect**.

1. Click **Web Application**.

   ![Application general setttings](/img/agentic-ai/screenshot-okta-app001.png)

1. Click **Next**.

1. Within **General Settings**, enter a name for your application
   in **App integration name**.

1. Click **Client Credentials**.

1. Click **Refresh Token**.

1. Click **Token Exchange**.

   ![Application general setttings](/img/agentic-ai/screenshot-okta-app002.png)

1. Set the **Sign-in redirect URIs** to that of the `oauth2c` default value:
   `http://localhost:9876/callback`

   ![Redirect URI setttings](/img/agentic-ai/screenshot-okta-app003.png)

1. Click **Allow everyone in your organization to access**.

1. Click **Enable immediate access with Federation Broker Mode**.

   ![Client access setttings](/img/agentic-ai/screenshot-okta-app004.png)

> 1. Click **Save**.

Reviewer: ````suggestion
1. Click **Save**.

1. Save the **Client ID** and **Client Secret** values listed on the general
   settings page. You need the secret values later to configure Vault:
 
     ```shell-session
   $ export \
       OKTA_CLIENT_ID=`<client_ID>`
       OKTA_CLIENT_SECRET=`<client_secret>`
   ```
````

### Get Client ID and Client Secret values

After saving your application, you land on the application general settings
page, where you can copy the Client ID and Client Secret values you need for
later validation of the agent workflow. Use the browser and a terminal session
to complete these steps.

1. Click the clipboard icon next to **Client ID** to copy the value.

1. In your terminal session, export the `OKTA_CLIENT_ID` environment variable
   and paste the Client ID as the value.

   ```shell-session
   $ export OKTA_CLIENT_ID=`<Client ID copied from previous step>`
   ```

1. Click the clipboard icon next to **Client Secrets** to copy the value.

1. In your terminal session, export the `OKTA_CLIENT_SECRET` environment
   variable and paste the Client Secret as the value.

   ```shell-session
   $ export OKTA_CLIENT_SECRET=`<Client Secret copied from previous step>`
   ```

>    ![Copy Client ID and Client Secret values](/img/agentic-ai/screenshot-okta-app005.png)

Reviewer: ```suggestion
```
Folded into the previous step.

As a general rule, we don't drill down to the level of showing folks how to copy values or calling out an explicit terminal session in how-to guides. They are less educational than tutorials and assume a basic level of understanding for common tools.

## Step 2: Configure an Authorization Server policy

Follow these steps to create an Authorization Server policy and associate
the policy with the application you created earlier. Use the browser and a

> terminal session to complete these steps.

Reviewer: ```suggestion
Create an authorization server policy and associate
the policy with your new OIDC application.
```

1. Click **Security**.

1. Click **API**.

1. Click **default** to select the default authorization server.

   ![Select default authorization server](/img/agentic-ai/screenshot-okta-app006.png)

1. From the **Issuer** value within **Settings**, copy the **Okta URL**
   value (the value in parentheses).

   ![Copy issuer URI](/img/agentic-ai/screenshot-okta-app007.png)

1. In your terminal session, export the `OKTA_URL` environment variable

>     and paste the Okta URL as the value.

Reviewer: ```suggestion
1. Save the issuer URL to an environment variable, `OKTA_URL`:
```

    ```shell-session
    $ export OKTA_URL=`<Okta URL copied from previous step>`
    ```

1. Return to the Okta UI in the browser and click **Access Policies**.

1. Click **Add Policy** to add new access policy.

   ![Add access policy](/img/agentic-ai/screenshot-okta-app008.png)

1. Enter a name into the **Name** field.

1. Add a description for the policy in the **Description** field.

1. For **Assign to**, click **The following clients** and then enter the
   name of the application you created.

   ![Assign access policy](/img/agentic-ai/screenshot-okta-app009.png)

1. Click **Create Policy**.

1. Click **Add rule**.

1. Enter a name for the rule into **Rule Name**.

1. Click to enable the **Authorization Code** checkbox.

1. Click to enable the **Device Authorization** checkbox.

1. Click **Advanced** to expand the options.

1. Click to enable the **Token Exchange** checkbox.

   ![Rule settings](/img/agentic-ai/screenshot-okta-app011.png)

1. Leave all other settings at their default values.

1. Click **Create rule**.

You configured the application, authorization server, and rules in Okta. You

> can now configure an authorization server scope.

Reviewer: ```suggestion
```

> ## Step 3: Create a Vault read scope

Reviewer: ```suggestion
## Step 3: Create an authorization server scope for Vault
```
Folded in the closing text from the previous step as it seemed a better explanation of what you're about to do in this step

Define a minimal scope for reading secrets from Vault and set it as the
default scope.

1. From the default authorization server **API** page, click **Scopes**.

1. Click **Add Scope**.

   ![Add scope](/img/agentic-ai/screenshot-okta-app012.png)

1. Enter `vault.read` into **Name**.

1. Enter a human-friendly phrase like `Vault Read` into **Display phrase**.

1. Click **Implicit**.

1. Click to check **Set as default scope**.

   ![Scope settings](/img/agentic-ai/screenshot-okta-app013.png)

1. Click **Create**.

## Step 4: Validate token retrieval

Use your application's Okta URL, client ID, and client secret values with the

> `oauth2c` tool to get a token from Okta as a validation step.

Reviewer: ```suggestion
Use `oauth2c` with your the Okta URL, client ID, and client secret values for
your OIDC app to get a token from Okta as a validation step.
```
Style correction: avoid possessives

> 1. Return to the terminal where you exported the environment variables.

Reviewer: ```suggestion
```

> 1. Request an example token with the oauth2c command-line tool.

Reviewer: ```suggestion
1. Request an example token with the `oauth2c` command-line tool:
```

    ```shell-session
    $ oauth2c "${OKTA_URL}" \
        --pkce                                             \
        --client-id "${OKTA_CLIENT_ID}"                    \
        --client-secret "${OKTA_CLIENT_SECRET}"            \
        --response-types code                              \
        --response-mode query                              \
        --grant-type authorization_code                    \
        --auth-method client_secret_basic                  \
        --no-browser
    ```

1. Note the URL under `Go to the following URL:` in the output. Visit that
   URL in a browser to complete the authentication flow. The workflow then

>    outputs more indicators of success.

Reviewer: ```suggestion
   URL in a browser to complete the authentication flow and verify the process
   succeeds.
```
Style correction: focus on what the user should do

   <CodeBlockConfig hideClipboard>

   ```plaintext
   ...snip...
    SUCCESS  Obtained authorization code
   ...snip...
   SUCCESS  Exchanged authorization code for access token
   ```

   </CodeBlockConfig>

   The final output string after `access_token` represents the JWT value,
   which validates both the workflow and token issuance from Okta.

1. Export the `access_token` value to the `JWT_ACCESS_TOKEN` environment

>    variable.

Reviewer: ```suggestion
   variable:
```

   ```shell-session
   $ export JWT_ACCESS_TOKEN=`<access_token value from previous step>`
   ```

> After you configure Vault, you can use the `JWT_ACCESS_TOKEN` value to

Reviewer: What does this mean? Part of validating the flow in the Vault setup is that you request a token. Why would they re-use this one?

authenticate your agent with Vault for secret access.

> </Tab>

Reviewer: I don't see a corresponding opening tab?

Author: Good catch, fixed.

> <Tab heading="Client credentials" group="client-credentials">

Reviewer: We shouldn't be tabbing the entire document. From what I can see, the actual steps ( create the app, define the policy, configure scope) are the same, it's just how you do them that changes. The tabs should be within each of those steps. For example:

```
Step 1: Create the OIDC service app

<token tab for step 1>
</tab>

<client credential tab for step 1>
</tab>
```

That way we're not replicating content throughout the doc. Also, you can't link to headings within a tab, so pulling the headings out from the tab lets folks bookmark specific steps or share them more easily.

<Note>

You must have an Okta account with the **NHI Authentication Tokens SKU**
enabled to use client credentials. This SKU is not available in the Okta free

> trial or Okta integrator accounts.

Reviewer: ```suggestion
Neither the free Okta trial, nor Okta integrator accounts, support authentication
with client credentials.
```
Style correction: write in active voice, avoid "this" as a pronoun

</Note>

> ## Step 1: Create the OIDC Service App

Reviewer: ```suggestion
```
We don't want headings within tabs if we can help it

You must first configure Okta with a Service App and custom authorization
server. You then define an authorization server policy and default scope

> for reading Vault secrets.

Reviewer: ```suggestion
```
Since this is basically the same intro, we can move it outside the tabs as shared content

   ![Redirect URI setttings](/img/agentic-ai/screenshot-okta-app003.png)

1. Click **Allow everyone in your organization to access**.

1. Click **Enable immediate access with Federation Broker Mode**.

   ![Client access setttings](/img/agentic-ai/screenshot-okta-app004.png)

1. Click **Save**.

### Get Client ID and Client Secret values

After saving your application, you land on the application general settings
page, where you can copy the Client ID and Client Secret values you need for
later validation of the agent workflow. Use the browser and a terminal session
to complete these steps.

1. Click the clipboard icon next to **Client ID** to copy the value.

1. In your terminal session, export the `OKTA_CLIENT_ID` environment variable
   and paste the Client ID as the value.

   ```shell-session
   $ export OKTA_CLIENT_ID=`<Client ID copied from previous step>`
   ```

1. Click the clipboard icon next to **Client Secrets** to copy the value.

1. In your terminal session, export the `OKTA_CLIENT_SECRET` environment
   variable and paste the Client Secret as the value.

   ```shell-session
   $ export OKTA_CLIENT_SECRET=`<Client Secret copied from previous step>`
   ```

>    ![Copy Client ID and Client Secret values](/img/agentic-ai/screenshot-okta-app005.png)

Reviewer: ```suggestion
```
This step is identical between the two workflows.  I suggest making it a conclusion to Step 1 instead of an explicitly numbered step so it doesn't need to be inside a tab

> ## Step 2: Configure an Authorization Server policy

Reviewer: ```suggestion
```
The instructions in this section should be a new tab under the step 2 block above

Follow these steps to create an Authorization Server policy and associate
the policy with the application you created earlier. Use the browser and a

> terminal session to complete these steps.

Reviewer: ```suggestion
```
